### ISSUE TITLE
[security-proxy] probable memory leak

### ISSUE THREAD

**Issue by pmauduit** [georchestra/georchestra#1069]
Some customers as well as the `sdi.georchestra.org` demo instance are encountering some downtimes on the tomcat hosting the security-proxy: The heap memory is drained from time to time, causing the tomcat to stop handling incoming connections.

**Comment by pmauduit** [comment]
An attempt to fix the issue:
https://github.com/pmauduit/georchestra/commit/02ff4ad70b171466227a21a41bb1d6de322c95f0

tested on a vm with capybara browser tests, below is a graph obtained by launching the tests in a loop during a whole night. No memory issue encountered, the testsuite was still running in the morning:

![proxycas-jmx](https://cloud.githubusercontent.com/assets/594335/10040523/bf6a7b62-61dc-11e5-8707-0045c414633e.png)

TODO: retest the same pattern without the previously mentioned fix.

**Comment by fvanderbiest** [comment]
:+1: :-)

**Comment by fvanderbiest** [comment]
Open question for everyone here: how long do you think should the session persist ? 1 hour, 4 hours, 8 hours, 24 hours or more ?

**Comment by pmauduit** [comment]
Note: the security-proxy already provides a default timeout set to 60minutes in the web.xml:
https://github.com/georchestra/georchestra/blob/15.06/security-proxy/src/main/webapp/WEB-INF/web.xml#L78

:+1: for 1 hour, which sounds reasonable to me.

**Comment by fphg** [comment]
assuming sessions timeout in 60mn on actual prod instances, I never heard anyone complain.
+1

**Comment by fvanderbiest** [comment]
> assuming sessions timeout in 60mn on actual prod instances

They don't.
If I understand correctly, they last (nearly) forever at the moment.

**Comment by fphg** [comment]
"the security-proxy already provides a default timeout set to 60minutes "
fake ?
Is that 68 years a single request session, or a login/logout session ?

**Comment by pmauduit** [comment]
68 years means that tomcat won't destroy the session for inactivity until 68 years period is reached. But maybe the line I removed from my commit affects the default specified in the web.xml.

Login/logout shall destroy the underlying session automatically, independantly of this timeout.

**Comment by pmauduit** [comment]
Same test without my patch:

![proxycas-jmx-used-nopatch](https://cloud.githubusercontent.com/assets/594335/10095418/fdcfed86-6369-11e5-8835-3a9fe0e7cb24.png)

Wrong track, or bad test conditions ?

**Comment by fvanderbiest** [comment]
Red line being the java heap used by tomcat-proxycas, we can see it increasing continuously on http://dev.geopicardie.fr (which is rarely used).

![capture du 2015-10-30 15 17 18](https://cloud.githubusercontent.com/assets/265319/10847929/9c693c7a-7f19-11e5-99b3-c4b92fa84e37.png)

(graph produced with graphite)

**Comment by pmauduit** [comment]
Some workarounds: http://stackoverflow.com/questions/3959122/memory-fully-utilized-by-java-concurrenthashmap-under-tomcat

**Comment by fvanderbiest** [comment]
@vdorut to tackle the issue

**Comment by pmauduit** [comment]
Almost reproduced locally:
![heap-consumption](https://cloud.githubusercontent.com/assets/594335/11748575/39cfae38-a029-11e5-902b-4455684f1f30.png)

Rennes-métropole encountered a tomcat-proxycas heap consumption 2 times in less than 2 days. We were able to grab another heap dump to analyze.

@vdorut did the work and found out that it was full of http sessions. What was surprising was the total amount of sessions compared to other dumps generated on other platforms (more than 54% as we observed before). One path is that since rennes-metropole is a closed platform, the monitoring probes need to be connected, we were able to find a large amount of sessions opened with principal (e.g. username) "c2c-monitoring".

**Comment by pmauduit** [comment]
Logs on production:

```
2015-12-11 05:33:55,057 [http-georchestra.mutu.local%2F127.0.1.1-8443-4] ERROR ina].[localhost].[/].[default] - Servlet.service() for servlet default threw exception
java.lang.OutOfMemoryError: GC overhead limit exceeded
        at org.springframework.ldap.core.SimpleCharStream.<init>(SimpleCharStream.java:261)
        at org.springframework.ldap.core.SimpleCharStream.<init>(SimpleCharStream.java:270)
        at org.springframework.ldap.core.DnParserImpl.<init>(DnParserImpl.java:192)
        at org.springframework.ldap.core.DefaultDnParserFactory.createDnParser(DefaultDnParserFactory.java:38)
        at org.springframework.ldap.core.LdapRdn.<init>(LdapRdn.java:56)
        at org.springframework.ldap.core.DistinguishedName.<init>(DistinguishedName.java:204)
        at org.springframework.ldap.core.DirContextAdapter.getNameInNamespace(DirContextAdapter.java:1265)
        at org.springframework.security.ldap.userdetails.LdapUserDetailsMapper.mapUserFromContext(LdapUserDetailsMapper.java:49)
        at org.springframework.security.ldap.authentication.AbstractLdapAuthenticationProvider.authenticate(AbstractLdapAuthenticationProvider.java:82)
        at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:156)
        at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:168)
        at org.georchestra.security.BasicAuthChallengeByUserAgent.doFilter(BasicAuthChallengeByUserAgent.java:58)
```

Logs on my dev env:

```
 java.lang.OutOfMemoryError: GC overhead limit exceeded
	at org.springframework.ldap.core.SimpleCharStream.<init>(SimpleCharStream.java:263)
	at org.springframework.ldap.core.SimpleCharStream.<init>(SimpleCharStream.java:270)
	at org.springframework.ldap.core.DnParserImpl.<init>(DnParserImpl.java:192)
	at org.springframework.ldap.core.DefaultDnParserFactory.createDnParser(DefaultDnParserFactory.java:38)
	at org.springframework.ldap.core.LdapRdn.<init>(LdapRdn.java:56)
	at org.springframework.ldap.core.DistinguishedName.<init>(DistinguishedName.java:204)
	at org.springframework.ldap.core.DirContextAdapter.<init>(DirContextAdapter.java:174)
	at org.springframework.ldap.core.support.DefaultDirObjectFactory.constructAdapterFromName(DefaultDirObjectFactory.java:174)
	at org.springframework.ldap.core.support.DefaultDirObjectFactory.getObjectInstance(DefaultDirObjectFactory.java:79)
	at javax.naming.spi.DirectoryManager.createObjectFromFactories(DirectoryManager.java:235)
	at javax.naming.spi.DirectoryManager.getObjectInstance(DirectoryManager.java:214)
	at com.sun.jndi.ldap.LdapSearchEnumeration.createItem(LdapSearchEnumeration.java:137)
	at com.sun.jndi.ldap.LdapNamingEnumeration.nextAux(LdapNamingEnumeration.java:274)
	at com.sun.jndi.ldap.LdapNamingEnumeration.nextImpl(LdapNamingEnumeration.java:254)
	at com.sun.jndi.ldap.LdapNamingEnumeration.next(LdapNamingEnumeration.java:202)
	at org.springframework.ldap.core.LdapTemplate.search(LdapTemplate.java:296)
	at org.springframework.ldap.core.LdapTemplate.search(LdapTemplate.java:259)
	at org.springframework.ldap.core.LdapTemplate.search(LdapTemplate.java:606)
	at org.springframework.ldap.core.LdapTemplate.search(LdapTemplate.java:524)
	at org.springframework.security.ldap.SpringSecurityLdapTemplate.searchForSingleAttributeValues(SpringSecurityLdapTemplate.java:173)
	at org.springframework.security.ldap.userdetails.DefaultLdapAuthoritiesPopulator.getGroupMembershipRoles(DefaultLdapAuthoritiesPopulator.java:215)
	at org.springframework.security.ldap.userdetails.DefaultLdapAuthoritiesPopulator.getGrantedAuthorities(DefaultLdapAuthoritiesPopulator.java:185)
	at org.springframework.security.ldap.authentication.LdapAuthenticationProvider.loadUserAuthorities(LdapAuthenticationProvider.java:197)
	at org.springframework.security.ldap.authentication.AbstractLdapAuthenticationProvider.authenticate(AbstractLdapAuthenticationProvider.java:82)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:156)
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:168)
	at org.georchestra.security.BasicAuthChallengeByUserAgent.doFilter(BasicAuthChallengeByUserAgent.java:95)

```

**Comment by fvanderbiest** [comment]
From http://stackoverflow.com/questions/5839359/java-lang-outofmemoryerror-gc-overhead-limit-exceeded : "For the record, we had the same problem today. We fixed it by using this option:

-XX:-UseConcMarkSweepGC
Apparently, this modified the strategy used for garbage collection, which made the issue disappear."

**Comment by fvanderbiest** [comment]
Might be worth a try...

**Comment by fphg** [comment]
Sure, thanks !
On Dec 13, 2015 8:15 AM, François Van Der Biest <notifications@github.com> wrote:Might be worth a try...

**Comment by landryb** [comment]
I think it was already advised by our documentation:
```
~/src/georchestra #git grep UseConcMarkSweepGC
doc/setup/tomcat.md:JAVA_OPTS="-Djava.awt.headless=true -XX:+UseConcMarkSweepGC"
doc/setup/tomcat.md:JAVA_OPTS="-Djava.awt.headless=true -XX:+UseConcMarkSweepGC"
doc/setup/tomcat.md:JAVA_OPTS="-Djava.awt.headless=true -Xmx128m -XX:+UseConcMarkSweepGC"
doc/setup/tomcat.md:            -XX:+UseConcMarkSweepGC"

```
and i have it in my playbooks:
```
~/georchestra-ansible #git grep UseConcMarkSweepGC
roles/tomcat/templates/config-georchestra.j2:JAVA_OPTS="-Djava.awt.headless=true -XX:+UseConcMarkSweepGC \
roles/tomcat/templates/config-geoserver.j2:JAVA_OPTS="-Djava.awt.headless=true -XX:+UseConcMarkSweepGC \
roles/tomcat/templates/config-proxycas.j2:JAVA_OPTS="-Djava.awt.headless=true -XX:+UseConcMarkSweepGC \

```

But.. isnt it a 'bandaid' instead of actually fixing the leaks/(aggressively) close the http sessions ? (yeah i know, easy to say...)

**Comment by landryb** [comment]
http://blog.sokolenko.me/2014/11/javavm-options-production.html also has plenty of sub-tweaks for the CMS GC...

**Comment by pmauduit** [comment]
:+1: with @landryb ; we could gain more knowledge on how the GC actually works and it would not be lost anyway, but it would be useless if it is a matter of uncollecteable objects

**Comment by fvanderbiest** [comment]
Of course, I was not saying "stop everything, I've got the best deal" ... in fact I'm not convinced at all this will change anything. I was just saying "it's worth a try".

Of course, +1 for going on with our research on the GC and "still referenced somewhere" / uncollected objects.

**Comment by fvanderbiest** [comment]
No proxy failure in ~ 3 weeks, with ```-XX:-UseConcMarkSweepGC``` and the d5acda1 patch.
Before that, tomcat-01 crashes happened every ~ 10 days.

Trying without the GC tweak today.

**Comment by fvanderbiest** [comment]
I think we fixed it with the d5acda1 patch.
Closing.
